certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+
+    require e.msg.sender != 0;  // safe assumption that prevents revert
+    require e.msg.value == 0;  // safe assumption that prevents revert
+
+    collectedAmt = collect@withrevert(e, userId, assetId);
+    assert !lastReverted;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// @param amt The given amount
+/// After give of amt, the splittable of the receiver should increase exactly by amt
+rule correctnessOfGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint128 splittableBefore; uint128 splittableAfter;
+
+    splittableBefore = splittable(e, receiver, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, receiver, assetId);
+
+    assert splittableAfter == splittableBefore + amt;
+}
+
+
+/// @notice The method give() gives funds from the user to the receiver.
+/// The splittable of any other user that is not receiver should not change
+rule splittableOfNonReceiverNotAffectedByGive() {
+    env e; uint256 userId; uint256 receiver; uint256 assetId; uint128 amt;
+    uint256 otherUser; uint128 splittableBefore; uint128 splittableAfter;
+
+    require otherUser != receiver;
+
+    splittableBefore = splittable(e, otherUser, assetId);
+    give(e, userId, receiver, assetId, amt);
+    splittableAfter = splittable(e, otherUser, assetId);
+
+    assert splittableAfter == splittableBefore;
+}
+
+
+/// @notice the method hashSplits() Calculates the hash of the list of splits receivers.
+/// We use boolean selector to operate on two different lists of splits receivers.
+/// We verify that two calculated hashes are the same only if they got exactly the same input
+/// We also verify that different inputs must generate different hashes
+rule correctnessOfHashSplits() {
+    env e; uint256 index; uint256 length1; uint256 length2;  
+    uint256 userId1; uint32 weight1; uint256 userId2; uint32 weight2;
+    bytes32 receiversHash1; bytes32 receiversHash2;
+
+    length1 = getCurrSplitsReceiverLocaLength(e, true);
+    length2 = getCurrSplitsReceiverLocaLength(e, false);
+
+    userId1, weight1 = getCurrSplitsReceiverLocalArr(e, true, index);
+    userId2, weight2 = getCurrSplitsReceiverLocalArr(e, false, index);
+
+    receiversHash1 = hashSplits(e, true);
+    receiversHash2 = hashSplits(e, false);
+
+    assert (receiversHash1 == receiversHash2) => (length1 == length2);
+    assert (receiversHash1 == receiversHash2) => (userId1 == userId2);
+    assert (receiversHash1 == receiversHash2) => (weight1 == weight2);
+
+    assert (length1 != length2) => (receiversHash1 != receiversHash2);
+    assert (length1 == length2) && (userId1 != userId2) => (receiversHash1 != receiversHash2);
+    assert (length1 == length2) && (weight1 != weight2) => (receiversHash1 != receiversHash2);
+
+    // assert (receiversHash1 == receiversHash2) <=> ((length1 == length2) && (userId1 == userId2) && (weight1 == weight2));
```

Comment:
It looks like that this assert is sufficient to prove what you want. The left to right side is what you've already asserted. You want to try to prove the right to left side.

Reply:
As we discussed, right to left is not correct, since only one of the two pairs is checked. Fixed it to be in left to right direction only.